### PR TITLE
feat!: remove the `Eq` type bound.

### LINKED ISSUE #1351: Remove `TypeBound::Eq`
See #1298. Eq should belong in a general system of interfaces, if such a thing comes to pass.

TypeBound to become about port connectivity (strong) or implicit or explicit copy/discard (weaker).

### PR DESCRIPTION
**Pull request by ss2165**
Closes #1351 

BREAKING CHANGE: Eq type bound removed. References to `Eq` in serialized HUGRs will be treated as `Copyable`.

### CODE REVIEW

File: hugr-core/src/hugr/serialize/test.rs [hunk 1]
```diff
@@ -79,7 +79,7 @@ macro_rules! include_schema {
             static ref $name: NamedSchema =
                 NamedSchema::new("$name", {
                     let schema_val: serde_json::Value = serde_json::from_str(include_str!(
-                        concat!("../../../../specification/schema/", $path, "_v2.json")
+                        concat!("../../../../specification/schema/", $path, "_v4.json")
```
Comment (aborgna-q) [inline hugr-core/src/hugr/serialize/test.rs:82]: `v3` is not even merged (#1352) and we are pushing for `v4` already? :P

File: hugr-core/src/types/poly_func.rs [hunk 2]
```diff
@@ -357,7 +357,7 @@ pub(crate) mod test {
     fn test_bound_covariance() -> Result<(), SignatureError> {
         decl_accepts_rejects_var(
             TypeBound::Copyable.into(),
-            &[TypeBound::Copyable.into(), TypeBound::Eq.into()],
+            &[TypeBound::Copyable.into(), TypeBound::Copyable.into()],
```
Comment (aborgna-q) [inline hugr-core/src/types/poly_func.rs:360]:
```suggestion
            &[TypeBound::Copyable.into()],
```

File: hugr-core/src/types/poly_func.rs [hunk 3]
```diff
@@ -366,7 +366,10 @@ pub(crate) mod test {
         };
         decl_accepts_rejects_var(
             list_of_tys(TypeBound::Copyable),
-            &[list_of_tys(TypeBound::Copyable), list_of_tys(TypeBound::Eq)],
+            &[
+                list_of_tys(TypeBound::Copyable),
+                list_of_tys(TypeBound::Copyable),
+            ],
```
Comment (aborgna-q) [inline hugr-core/src/types/poly_func.rs:372]:
```suggestion
            &[list_of_tys(TypeBound::Copyable)],
```

File: hugr-py/src/hugr/serialization/tys.py [hunk 4]
```diff
@@ -381,18 +381,17 @@ def deserialize(self) -> tys.PolyFuncType:
 
 
 class TypeBound(Enum):
-    Eq = "E"
     Copyable = "C"
     Any = "A"
 
     @staticmethod
     def join(*bs: TypeBound) -> TypeBound:
```
Comment (aborgna-q) [inline hugr-py/src/hugr/serialization/tys.py:388]: Not covered by tests?

File: specification/hugr.md [hunk 5]
```diff
@@ -785,24 +785,20 @@ indices after the list of node indices?
 
 ## Type System
 
-There are three classes of type: `AnyType` $\supset$ `CopyableType` $\supset$ `EqType`. Types in these classes are distinguished by the operations possible on (runtime) values of those types:
+There are two classes of type: `AnyType` $\supset$ `CopyableType`. Types in these
+classes are distinguished by whether the runtime values of those types can be implicitly
+copied or discard (multiple or 0 links from on output port respectively):
```
Comment (aborgna-q) [inline specification/hugr.md:790]:
```suggestion
copied or discarded (multiple or 0 links from on output port respectively):
```

**Review by aborgna-q — COMMENTED** [review]
The schema version update is missing things (updating the version tags in the generated jsons, both in the hugr def and in hugr-py). This will probably fail in ci due to the mismatched schema file version vs `hugr.get_serialisation_version()`.

Is this PR meant to complement #1352?

File: hugr-core/src/hugr/serialize/test.rs (hunk 1 above)
Comment (ss2165) [reply hugr-core/src/hugr/serialize/test.rs:82]: yes this is assuming 1352 goes in first

File: hugr-py/src/hugr/serialization/tys.py (hunk 4 above)
Comment (ss2165) [reply hugr-py/src/hugr/serialization/tys.py:388]: probably pulled in from guppy, new issue : #1365

File: hugr-core/src/hugr/serialize/test.rs (hunk 1 above)
Comment (ss2165) [reply hugr-core/src/hugr/serialize/test.rs:82]: can we just include this in the v3 change?

Comment (ss2165) [reply hugr-core/src/hugr/serialize/test.rs]: have reverted schema changes pending #1352

Comment (aborgna-q) [reply hugr-core/src/hugr/serialize/test.rs]: 👍 
If this goes in the same release we should bundle do a single version bump.

**Review by aborgna-q — APPROVED** [review]

**Pull request merged**